bam_sort.c

```diff
@@ -1591,13 +1669,26 @@ typedef struct {
 
 static inline int heap_add_read(heap1_t *heap, int nfiles, samFile **fp,
                                 int num_in_mem, buf_region *in_mem,
-                                bam1_tag *buf, uint64_t *idx, sam_hdr_t *hout) {
+                                bam1_tag *buf, template_coordinate_keys_t *keys,
+                                uint64_t *idx, sam_hdr_t *hout) {
     int i = heap->i, res;
     if (i < nfiles) { // read from file
         res = sam_read1(fp[i], hout, heap->entry.bam_record);
+        if (res >= 0 && g_sam_order == TemplateCoordinate) { // file read OK and TemplateCoordinate order
+            if (keys->n >= keys->m * keys->buffer_size) res = template_coordinate_keys_realloc(keys, keys->n + 1); // need more memory
+            if (res >= 0) {
+                template_coordinate_key_t *key = template_coordinate_keys_get(keys, keys->n); // get the next key to use
+                heap->entry.u.key = template_coordinate_key(heap->entry.bam_record, key, hout); // update the key
+                if (heap->entry.u.key == NULL) res = -1; // key could not be created, error out
+                else keys->n++; // key created OK, increment the number of keys
+            }
+        }
```

Comment:
It looks like you'll use much more memory than needed here, as you're adding a new key for every single record read back in.  It should instead be possible to recycle them in the same way as the bam_record structs.  The easiest way to do this would be to use entry `keys->n + i` instead of just `keys->n`, without incrementing `keys->n`.

Reply:
Are you suggesting to use the additional keys between `[k->n,k->m)` ?  I.e. the keys that have not been used but have been allocated?  I think that's what I'm doing.

We need enough memory to store the existing in-memory keys (already stored in`[0, k->n)`), and any additional keys that need to be made from bam records read from the file.  For the latter keys, why not add them into `keys` right after the existing in-memory keys?  

Or you saying I don't need to re-allocate the keys at all, and I can somehow use keys within `[0, k->n)`?  Or just assume `heap->entry.u.key` is not `NULL`?  

My apologies as I don't understand your suggestion.

Reply:
Done!